Port some Hermes-3 finite volume operators

ZedThree opened the pull request:
Includes #3196 

Ports the following from [Hermes-3 `fci-more-fixes-update-more`](https://github.com/boutproject/hermes-3/tree/fci-more-fixes-update-more):

- `FV::Div_par_mod`
- `FV::Div_par_fvv`
- `Div_par_K_Grad_par_mod` (not strictly FV!)
- `FV::Superbee`

Plus adds MMS tests for these and the existing FV operators:

- `FV::Div_par`
- `FV::Div_par_K_Grad_par`

with all the available slope limiters, as appropriate.

----

Because the slope limiters kind of necessarily introduce inaccuracies, I've reduced the expected order of the operators, with `Upwind` having order 1 and the rest 1.5. This feels like a bit of a fudge, but it's not clear to me how to treat this better.

----

The ported operators will obviously conflict with the Hermes-3 ones. So 2/3 things to do:

- Remove the ported ones from Hermes-3 when updating the bundled BOUT++
- Move all the operators here into `bout::FV::` namespace
- Move all the operators there into `hermes::FV::` namespace

ZedThree committed:
tests: Ruff fixes for FCI runtest

ZedThree committed:
tests: Expand FCI MMS test to `Grad2_par2`

ZedThree committed:
tests: Generalise FCI MMS test to allow for more cases

ZedThree committed:
tests: Add test for FCI `Div_par`

ZedThree committed:
tests: Make MMS script update input in-place

ZedThree committed:
tests: Add test for FCI `Div_par_K_Grad_par`

ZedThree committed:
Many small fixes for FCI

- Include all relevant headers, remove unused ones, add some forward
  declarations
- Make data `private` instead of `protected`
- Add getter instead of `const` member
- Change member reference to pointer
- Move ctor to .cxx file
- Use `std::array` instead of C array
- Bunch of other small clang-tidy fixes

ZedThree committed:
tests: Fix for 3D metric in FCI test

The Div_par operators use parallel slices of B -- with 2D metrics,
these are just the field itself, in 3D we need the actual
slices. While `Coordinates::geometry` does communicate the fields, it
puts off calculating the parallel slices because that needs the fully
constructed `Coordinates`.

Upcoming changes should fix this and remove the need to explicitly
communicate `Coordinates` members.

ZedThree committed:
tests: Add test for FCI `Laplace_par`

ZedThree committed:
Reduce duplication in FCI mms script

ZedThree committed:
Remove circular include

ZedThree committed:
tests: Add cases for FCI interpolation methods

ZedThree committed:
tests: Increase nx for hermitespline interpolation boundary problem

ZedThree committed:
tests: Add monotonichermitespline, decrease resolution scan

ZedThree committed:
tests: Small refactor of FCI tests

ZedThree committed:
Apply black changes

ZedThree committed:
Apply clang-format changes

ZedThree committed:
Apply clang-tidy fixes to FCI tests

ZedThree committed:
Fix comment formatting

ZedThree committed:
Apply suggestion from @dschwoerer

Co-authored-by: David Bold <dschwoerer@users.noreply.github.com>

ZedThree committed:
Port `FV::Div_par_mod` from Hermes-3

ZedThree committed:
Port `FV::Div_par_fvv` from Hermes-3

ZedThree committed:
Port `Div_par_K_Grad_par_mod` from Hermes-3

ZedThree committed:
Add MMS tests for finite volume operators

ZedThree committed:
tests: Fix typo that made tests always pass

ZedThree committed:
tests: Remove monotonichermitespline FCI case

ZedThree committed:
Add tests for all FV slope limiters

ZedThree committed:
Port `Superbee` finite volume limiter from Hermes-3

Includes bug fix:

```diff
-    BoutReal gL = n.c - n.L;
-    BoutReal gR = n.R - n.c;
+    BoutReal gL = n.c - n.m;
+    BoutReal gR = n.p - n.c;
```

ZedThree committed:
Fix clang-tidy warnings for fv_ops

ZedThree committed:
Apply ruff format

ZedThree committed:
tests: Communicate some coordinates fields for 3D metrics

dschwoerer commented on tests/MMS/spatial/finite-volume/fv_mms.cxx:
```diff
@@ -42,6 +42,15 @@ int main(int argc, char** argv) {
+    // Div_par operators require B parallel slices:
+    // Coordinates::geometry doesn't ensure this (yet)
+    auto& Bxy = mesh->getCoordinates()->Bxy;
+    auto& J = mesh->getCoordinates()->J;
```
Note that communicating J is likely wrong for any realistic grid. You need #3197 
For a simple slab test, that may well work, but not for real grids.

dschwoerer replied:
The parallel slices can have a completely different geometry. Thus one needs J from the traced parallel slices. The communication interpolates J, which works for lab-based quantities, but not for mesh based quantities.

One solution is to also write out the parallel metric components in zoidberg, and read them in in BOUT++, and prevent them from being overwritten:
https://github.com/boutproject/zoidberg/blob/more-6/zoidberg/zoidberg.py#L421
https://github.com/boutproject/BOUT-dev/blob/f3dwy/src/mesh/parallel/fci.cxx#L134

See also:
https://bout.zulipchat.com/#narrow/channel/535664-BSTING/topic/Connection.20between.20J.20and.20B.20in.20FCI/with/554530388

ZedThree committed:
Include all headers

ZedThree committed:
Add `rtol`, `atol` to `MonotonicHermiteSpline`

This allows the user to control the clipping, allowing some overshoot

Also restore the FCI MMS test with this interpolator

ZedThree committed:
Div_par_fvv: FCI change numerical dissipation

Damp gradients of velocity rather than gradients of momentum

ZedThree committed:
Set expected order of `FV::Div_par_fvv` to 1

ZedThree committed:
Add missing header

ZedThree committed:
Communicate some more coordinate quantities for FCI MMS test

ZedThree committed:
Guard against non-thread safe PETSc call

ZedThree committed:
Remove some unused headers

mrhardman commented:
@ZedThree Is there a possibility that this PR could also cover the operators and MMS tests of operators featured https://github.com/boutproject/hermes-3/pull/285 and https://github.com/boutproject/hermes-3/pull/327? The latter PR also has model-level tests that could not be ported to BOUT-dev. I imagine that the functions in https://github.com/boutproject/hermes-3/blob/nonorthogonal-debug/tests/mms/perpendicular_laplacian.py could be factored into https://github.com/boutproject/boutdata/blob/master/src/boutdata/mms.py.

mrhardman commented:
> Yes, certainly possible, will investigate!

Thank you! I made a start on trying to make the existing tests in my latest PR https://github.com/boutproject/hermes-3/pull/327 more similar to `boutdata.mms` here https://github.com/boutproject/hermes-3/pull/327/commits/32c8536f98a144329b1fe56515fd3e3d7c8cc792. It likely better to view the files directly. I note that the fact that the operators use an optional argument for `metric` by convention tripped me up as I tried to propagate my `GeneralMetric` to all functions, as it was possible to mistakenly not pass a consistent metric to all differential operators.

dschwoerer reviewed (dismissed):
Looks good!

I did not notice that this PR was still pending, and rebased it on top of the recent changes, including #3320 and #3334.

Do you prefer resolving the conflicts here and merge this, or use #3335 instead?

bendudson commented:
Replaced by #3335
